pkg/raft/raft_test.go

```diff
@@ -2572,22 +2573,49 @@ func TestLeaderAppResp(t *testing.T) {
 		reject bool
 		// progress
 		wmatch uint64
-		wnext  uint64
-		// message
-		wmsgNum    int
-		windex     uint64
+		// log index of the next entry to send to this follower.
+		wnext uint64
+		// messageCount the leader sends out
+		wmsgNum int
+		// prevLogIndex in MsgApp from leader to followers
+		windex uint64
+		// leader's commit index
 		wcommitted uint64
+		// storage access counts for getting term number
+		ctStgTerm int
 	}{
-		{2, true, 0, 4, 0, 0, 0}, // stale resp; no replies
-		{6, true, 0, 4, 0, 0, 0}, // stale resp; no replies
-		{3, true, 0, 3, 1, 2, 0}, // denied resp; leader does not commit; decrease next and send probing msg
-
-		{4, false, 4, 7, 1, 4, 4}, // accept resp; leader commits; broadcast with commit index
-		// Follower is StateProbing at 4, it sends MsgAppResp for 4, and is moved to
+		{2, true, 0, 4, 0, 0, 0, 1}, // stale resp; no replies
+		{6, true, 0, 4, 0, 0, 0, 1}, // stale resp; no replies
+
+		// denied resp; leader does not commit; decrease next and send probing msg
+		// An additional term storage access is involved for an entry
+		// that's already persisted since we are probing backwards
+		{3, true, 0, 3, 1, 2, 0, 2},
+
+		// Follower2 responds to leader, indicating log index2 is replicated.
+		// Leader tries to commit, but commit index doesn't advance since the index is from a previous term
+		// We hit maybeCommit() and do term check by getting the term number from storage
+		{2, false, 2, 7, 1, 2, 0, 3},
+
+		// NB: For the following tests, we are skipping the MsgAppResp for the first 3 entries, by directly processing MsgAppResp for later entries
+		//
+		// Follower2 is StateProbing at 4, it sends MsgAppResp for 4, and is moved to
 		// StateReplicate and as many entries as possible are sent to it (5, 6).
 		// Correspondingly the Next is then 7 (entry 7 does not exist, indicating
 		// the follower will be up to date should it process the emitted MsgApp).
-		{4, false, 4, 7, 1, 4, 4},
+		// accept resp; leader commits; broadcast with commit index
+		{4, false, 4, 7, 1, 4, 4, 1},
+
+		// Follower2 says term2, index5 is already replicated
+		// The leader broadcasts to followers saying the entry{term:2, index:5} has been committed, with commit index 5
+		// The leader sends entry{term:2, index:6} to follower2
```

Comment:
Is this really what happens? I think The leader just sends a MsgApp with the updated commit index, but no entries. All the entries are already in-flight.